src/java.base/share/classes/java/text/MessageFormat.java

```diff
@@ -562,60 +654,81 @@ public String toPattern() {
         for (int i = 0; i <= maxOffset; ++i) {
             copyAndFixQuotes(pattern, lastOffset, offsets[i], result);
             lastOffset = offsets[i];
-            result.append('{').append(argumentNumbers[i]);
-            Format fmt = formats[i];
-            if (fmt == null) {
-                // do nothing, string format
-            } else if (fmt instanceof NumberFormat) {
-                if (fmt.equals(NumberFormat.getInstance(locale))) {
-                    result.append(",number");
-                } else if (fmt.equals(NumberFormat.getCurrencyInstance(locale))) {
-                    result.append(",number,currency");
-                } else if (fmt.equals(NumberFormat.getPercentInstance(locale))) {
-                    result.append(",number,percent");
-                } else if (fmt.equals(NumberFormat.getIntegerInstance(locale))) {
-                    result.append(",number,integer");
-                } else {
-                    if (fmt instanceof DecimalFormat) {
-                        result.append(",number,").append(((DecimalFormat)fmt).toPattern());
-                    } else if (fmt instanceof ChoiceFormat) {
-                        result.append(",choice,").append(((ChoiceFormat)fmt).toPattern());
-                    } else {
-                        // UNKNOWN
-                    }
+            result.append('{')
+                    .append(argumentNumbers[i])
+                    .append(patternFromFormat(formats[i]))
+                    .append('}');
+        }
+        copyAndFixQuotes(pattern, lastOffset, pattern.length(), result);
+        return result.toString();
+    }
+
+    /**
+     * This method converts a Format into a {@code FormatType} and {@code
+     * FormatStyle}, if applicable. For each Format, this method will
+     * first check against the pre-defined styles established in the
+     * {@link ##patterns patterns section}. If a Format does not match to a
+     * pre-defined style, it will provide the {@code SubformatPattern}, if the Format
+     * class can provide one. The following subformats do not provide a {@code
+     * SubformatPattern}: {@link CompactNumberFormat}, {@link ListFormat}, and {@link
+     * java.time.format.DateTimeFormatter}. In addition, {@code DateTimeFormatter}
+     * does not implement {@code equals()}, and thus cannot be represented as a
+     * pattern. Any "default"/"medium" styles are omitted according to the specification.
```

Comment:
We discussed this separately, but will go over again here in case others are interested. Since DTF does not implement `equals()`, even if we implement `equals()` for the `ClassicFormat`, we would basically still need to implement our own `equals()` for a DTF to effectively compare the ClassicFormats.

I had also mentioned that we could reference check the pre-defined DTFs, but this won't work either actually. This is because we cannot reference check the DTFs, but rather the Format adapted DTFs, which mean they are now new ClassicFormats every time since we don't store them.

Reply:
Thanks. I think it is fine.

Reply:
The comment "does not implement equals and thus cannot be represented as a pattern" doesn't explain why it can be represented as a pattern. Can that be explained better?